src/track/track_gps_l2cm.c

```diff
@@ -0,0 +1,614 @@
+ * interrupt arrives, the L2C tracker gets the first chance to change the
+ * integration length. It sets it to be
+ * L2CM_TRACK_LONG_STARTUP_CYCLE_INTERVAL_CHIPS chips. However, NAP continues
+ * doing integration and still uses the previous integration length, which is
+ * L2CM_TRACK_SHORT_CYCLE_INTERVAL_CHIPS chips long. At the next interrupt
+ * FW sets the integration length to be L2CM_TRACK_SHORT_CYCLE_INTERVAL_CHIPS.
+ *
+ * \param channel_info INfo associated with a tracker channel
+ * \param common_data Common tracking channel data
+ * \param data L2C specific tracking channel data
+ */
+static void handle_tracker_startup(const tracker_channel_info_t *channel_info,
+                                   tracker_common_data_t *common_data,
+                                   gps_l2cm_tracker_data_t *data)
+{
+  corr_t cs[3];
+
+  switch (data->startup) {
+  case 2:
+    tracker_correlations_read(channel_info->context, data->cs,
+                              &common_data->sample_count,
+                              &common_data->code_phase_early,
+                              &common_data->carrier_phase);
+    alias_detect_first(&data->alias_detect, data->cs[1].I, data->cs[1].Q);
+
+    tracker_retune(channel_info->context, common_data->carrier_freq,
+                   common_data->code_phase_rate,
+                   L2CM_TRACK_LONG_STARTUP_CYCLE_INTERVAL_CHIPS);
+    break;
+
+  case 1:
+    tracker_correlations_read(channel_info->context, cs,
+                              &common_data->sample_count,
+                              &common_data->code_phase_early,
+                              &common_data->carrier_phase);
+    /* Accumulate two short cycle correlations */
+    for(int i = 0; i < 3; i++) {
+      data->cs[i].I += cs[i].I;
+      data->cs[i].Q += cs[i].Q;
+    }
+    tracker_retune(channel_info->context, common_data->carrier_freq,
+                   common_data->code_phase_rate,
+                   L2CM_TRACK_SHORT_CYCLE_INTERVAL_CHIPS);
+    break;
+
+  default:
+    assert(0);
+    break;
+  }
+}
+
+static void tracker_gps_l2cm_update(const tracker_channel_info_t *channel_info,
+                                    tracker_common_data_t *common_data,
+                                    tracker_data_t *tracker_data)
+{
+  gps_l2cm_tracker_data_t *data = tracker_data;
+
+  if (data->startup) {
+    /* Handle start-up case, when we have 1+1+18 integration cycles */
+    handle_tracker_startup(channel_info, common_data, data);
+    if (1 == data->startup) {
+      /* The start-up phase is over. Continue with the normal
+         sequence of short & long cycles. The next cycle
+         is going to be long. */
+      data->short_cycle = false;
+    }
+    if (data->startup) {
+      data->startup--;
+    }
+    return;
+  }
+
+  /* Read early ([0]), prompt ([1]) and late ([2]) correlations. */
+  if (data->short_cycle) {
+    /* The throw away data. Not needed for the short cycle.
```

Comment:
The short cycle on L2C is set to ~300 us at the moment (L2CM_TRACK_SHORT_CYCLE_INTERVAL_CHIPS). The long cycle is ~19700 us. The TOW is only updated after the long cycle is over. So I do not call  tracker_tow_update() after the short cycle, because tracker_tow_update expects full ms updates to TOW. Since TOW is not updated on the short cycle, I do not update common_data struct either to avoid the inconsistency.